### ISSUE TITLE
Explore alternative storage solutions and data sharing across threads

### ISSUE THREAD

**Issue by mohdsayed** [GoogleChromeLabs/ps-analysis-tool#403]
## Task

Epic: https://github.com/GoogleChromeLabs/ps-analysis-tool/issues/380

### Description

This issue involves research work to find alternative storage solutions to enable data sharing across threads, and if we even need to use additional storage for sharing data.

**Background**

Currently, we are using `chrome.storage.local` to store all cookie data in the service worker. This allows us to share the data with the DevTools and Popup, as it is available in the context of every thread. To prevent data loss and ensure single concurrency, we push all updates to the local storage in a promise queue.

We have two main reasons for using this approach:

1. To prevent data loss, as local storage does not have all ACID properties and multiple concurrent requests can override each other.
2. To avoid crashes or freezing of the extension, which occurs when we do not use a promise queue.

**Problem**

As the extension grows and with the addition of CDP in the last version, we have recently added more items to the promise queue on certain CDP events. This has resulted in noticeable delays in rendering the cookie data in the table.

**Initial Ideas**

1. Find an alternative solution for storage APIs. For example, consider using IndexedDB in each thread and synchronize the data among threads using message passing at regular intervals.

![image](https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/6297436/e7e6c712-798f-4013-9cd9-db8586099265)

2. Alternatively, store the cookie data in cache memory and avoid using any storage for cookie data. Share the data across threads using message passing. Note that we would still use local storage for storing preferences and settings.

![image](https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/6297436/d47f47d6-989d-4345-9daa-53b3cd9c685e)

**Acceptance Criteria:**
- [x] Create a POC for the proposed approach.
- [x] Measure the performance of the POC through multiple tests.

Related https://github.com/GoogleChromeLabs/ps-analysis-tool/issues/380

**Comment by gagan0123** [comment]
## Benchmarking Results of Different Chrome Storage Methods

Performed benchmarking tests to assess the performance of different storage methods in Chrome. The objective was to compare the time efficiency of `chrome.storage.local`, `chrome.storage.session`, `chrome.storage.sync`, and `IndexedDB` in a Chrome extension environment. The test consisted of storing and retrieving 10,000 random key-value pairs for each storage method and measuring the duration of these operations.

### Methodology

The benchmarking was executed through a custom Chrome extension. The following JavaScript code snippet illustrates the core logic used for the performance measurement:

<details>
  <summary>Click to view the benchmarking code</summary>
  
  ```jsx
async function measureTime(storageType) {
    let startTime = performance.now();

    for (let i = 0; i < 10000; i++) {
        let key = `key${i}`;
        let value = Math.random().toString(36).substring(7);

        await new Promise(resolve => storageType.set({ [key]: value }, resolve));
        await new Promise(resolve => storageType.get(key, resolve));
    }

    let endTime = performance.now();
    return endTime - startTime;
}

// Measure time consumed in ms
const timeTakenLocal = await measureTime(chrome.storage.local);
const timeTakenSession = await measureTime(chrome.storage.session);
const timeTakenSync = await measureTime(chrome.storage.sync);
// IndexedDB measurement follows a similar pattern.

```
</details>



### Findings

The table below presents the results of the benchmarking for setting and retrieving 10,000 random key-value pairs. Time measured in milliseconds (ms):

| **Storage Type** | **Average (ms)**  |
|------------------|--------------|
| **local**        | **23625.7698** |
| **session**      | **1524.0657**  |
| **sync**         | **1788.31**    |
| **indexedDB**    | **1237.8388**  |


<details>
  <summary>Detailed results</summary>
  
  | Storage Type | Average (ms) | Test 1 | Test 2 | Test 3 | Test 4 | Test 5 | Test 6 | Test 7 | Test 8 | Test 9 | Test 10 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| local | 23625.7698 | 14302.1 | 26428.3 | 13674.9 | 27112.3 | 27794.2 | 27285.1 | 27407.1 | 27758.099 | 27148 | 17347.599 |
| session | 1524.0657 | 1618.7 | 1402.59 | 1424.5 | 1501.59 | 1569.699 | 1431.09 | 1594.59 | 1445.2 | 1427.099 | 1825.599 |
| sync | 1788.31 | 1898.4 | 1615.8 | 1634.3 | 1624.5 | 1795.4 | 1684.4 | 1790.3 | 1826.8 | 1846.8 | 2166.4 |
| indexedDB | 1237.8388 | 1264.8 | 1200.59 | 1097 | 1220.8 | 1267.199 | 1121 | 1384.9 | 1174 | 1159.8 | 1488.299 |
</details>

### Analysis

- **Local Storage**: Took the longest time. This might be due to its larger storage capacity and persistence.
- **Session Storage**: Significantly faster than local storage, potentially due to its transient nature and smaller storage capacity.
- **Sync Storage**: Performed similarly to session storage, although slightly slower.
- **IndexedDB**: Fastest among the tested methods.

### Conclusion

The benchmarking results clearly show that **`chrome.storage.local`** is the slowest among the tested storage methods. Given this, we should switch from using **`chrome.storage.local`** to more efficient storage options.

**Comment by maitreyie-chavan** [comment]
Addressed and released as a part of [v0.5](https://github.com/GoogleChromeLabs/ps-analysis-tool/releases/tag/v0.5.0)
